### ISSUE TITLE
Product repository "get" method returns error if called more than 1000 times

### ISSUE THREAD

**Issue by dankocherga** [magento/magento2#12355]
### Preconditions
1. Magento 2.2.1

### Steps to reproduce
1. Setup a database with more than 1000 different products
2. Call repository `get` method with different SKUs

### Expected result
1. Each call returns a product by SKU

### Actual result
After some calls:
```
PHP Notice:  Undefined offset: <SKU_STRING_XYZ> in /vendor/magento/module-catalog/Model/ProductRepository.php on line 254
```

Seems like the problem is in the following:
* `cacheProduct` [method is called](https://github.com/magento/magento2/blob/2.2.1/app/code/Magento/Catalog/Model/ProductRepository.php#L249) for each found item.
* If `cacheLimit` field is set (1000 by default) and number if cached items exceeds this limit, then cache array is [sliced by a half](https://github.com/magento/magento2/blob/2.2.1/app/code/Magento/Catalog/Model/ProductRepository.php#L312-L316).
* Then repository method [gets item from cache](https://github.com/magento/magento2/blob/2.2.1/app/code/Magento/Catalog/Model/ProductRepository.php#L254) and it might happen that this key was in sliced half. So it results in error.

**Comment by magento-engcom-team** [comment]
@dankocherga, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. If you'd like to update it, please reopen the issue.
We tested the issue on 2.3.0-dev, 2.1.10, 2.2.1

**Comment by orlangur** [comment]
@dankocherga I believe this was fixed in https://github.com/magento/magento2/commit/3b50751687f873a645b914194228e0006fa1cb1b which will be released in 2.2.2.

**Comment by fsw** [comment]
Actually 3b50751 is just fixing 50% cases. cacheProduct is still clearing half the cache AFTER setting the product so there is 50% chance that the product we just set will be deleted just after and cause this error. I think 1270ab9cd514f189a1c05d4ee1f90bed95f5bdf8 is fixing this permanently by changing get() method flow.
